cmd/rig/cmd/dev/host.go

```diff
@@ -0,0 +1,418 @@
+				Content: bs,
+			},
+		},
+	})
+
+	var interfaces []*capsule.Interface
+	for _, capIf := range cfg.GetNetwork().GetCapsuleInterfaces() {
+		interfaces = append(interfaces, &capsule.Interface{
+			Port: capIf.Port,
+			Name: fmt.Sprintf("forward-%d", capIf.Port),
+		})
+	}
+
+	if cfg.GetNetwork().GetTunnelPort() != 0 {
+		interfaces = append(interfaces, &capsule.Interface{
+			Port: cfg.GetNetwork().GetTunnelPort(),
+			Name: "host-tunnel",
+		})
+	}
+
+	changes = append(changes, &capsule.Change{
+		Field: &capsule.Change_Network{
+			Network: &capsule.Network{
+				Interfaces: interfaces,
+			},
+		},
+	})
+
+	changes = append(changes, &capsule.Change{Field: &capsule.Change_AddImage_{
+		AddImage: &capsule.Change_AddImage{
+			Image: "ghcr.io/rigdev/rig-proxy:dev",
+		},
+	}})
+
+	revision, rolloutID, err := capsule_cmd.Deploy(ctx, c.Rig, c.Scope, capsuleName, changes, true, false, 0)
+	if err != nil {
+		return err
+	}
+
+	if err := capsule_cmd.WaitForRollout(ctx, c.Rig, c.Scope, capsuleName, revision, rolloutID); err != nil {
+		return err
+	}
+
+	instanceID := ""
+
+	if instanceID == "" {
+		if instanceID, err = capsule_cmd.GetCapsuleInstance(ctx, c.Rig, c.Scope, capsuleName); err != nil {
+			return err
+		}
+	}
+
+	capInterfaces := map[uint32]*platformv1.CapsuleInterface{}
+	for _, capIf := range cfg.GetNetwork().GetCapsuleInterfaces() {
+		capInterfaces[capIf.GetPort()] = capIf
+	}
+
+	l, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		return err
+	}
+
+	gc, err := grpc.NewClient(l.Addr().String(), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return err
+	}
+
+	ls := map[uint32]hostListener{}
+	for _, hostIf := range cfg.GetNetwork().GetHostInterfaces() {
+		l, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:%d", hostIf.GetPort()))
+		if err != nil {
+			return err
+		}
+
+		_, portStr, err := net.SplitHostPort(l.Addr().String())
+		if err != nil {
+			return err
+		}
+
+		port, err := strconv.ParseUint(portStr, 10, 32)
+		if err != nil {
+			return err
+		}
+
+		ls[uint32(port)] = hostListener{
+			cfg:      hostIf,
+			listener: l,
+		}
+	}
+
+	tunnelClient := api_tunnel.NewServiceClient(gc)
+	go func() {
+		for {
+
+			tunnelStream, err := tunnelClient.Tunnel(ctx)
+			if err != nil {
+				fmt.Println("[rig] error establishing tunnel: ", err)
+				time.Sleep(1 * time.Second)
+				continue
+			}
+
+			rt := &ReverseTunnel{
+				tunnelStream:   tunnelStream,
+				capInterfaces:  capInterfaces,
+				tunnels:        map[uint64]*tunnel.Buffer{},
+				hostInterfaces: ls,
+			}
+
+			if err := rt.Run(ctx); err != nil {
+				fmt.Println("[rig] err processing tunnel: ", err)
+			}
+
+			time.Sleep(1 * time.Second)
+		}
+	}()
+
+	return capsule_cmd.PortForwardOnListener(
+		ctx, c.Rig, c.Scope, capsuleName, instanceID, l, cfg.GetNetwork().GetTunnelPort(), true)
+}
+
+type hostListener struct {
+	cfg      *platformv1.HostInterface
+	listener net.Listener
+}
+
+type ReverseTunnel struct {
```

Comment:
Naming wise, isn't it just a tunnel? Considering it handles both host and capsule interfaces it doesn't have a preferred direction?

Reply:
It is :)